cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go

```diff
@@ -42,15 +43,40 @@ type ClusterSnapshot interface {
 	// IsPVCUsedByPods returns if the pvc is used by any pod, key = <namespace>/<pvc_name>
 	IsPVCUsedByPods(key string) bool
 
-	// Fork creates a fork of snapshot state. All modifications can later be reverted to moment of forking via Revert()
-	// Forking already forked snapshot is not allowed and will result with an error.
-	Fork() error
+	// Fork creates a fork of snapshot state. All modifications can later be reverted to moment of forking via Revert().
+	// Use WithForkedSnapshot() helper function instead if possible.
+	Fork()
 	// Revert reverts snapshot state to moment of forking.
 	Revert() error
 	// Commit commits changes done after forking.
 	Commit() error
-	// Clear reset cluster snapshot to empty, unforked state
+	// Clear reset cluster snapshot to empty, unforked state.
 	Clear()
 }
 
 var errNodeNotFound = errors.New("node not found")
+
+// WithForkedSnapshot is a helper function for snapshot that makes sure all Fork() calls are closed with Commit() or Revert() calls.
+// Error for this function means that cluster snapshot is in broken state.
+func WithForkedSnapshot(snapshot ClusterSnapshot, f func() error) error {
```

Comment:
I don't really like using error returned by passing function as a signal to commit/revert. 

I think the decision to commit or revert depends more on what the forked snapshot is used for (scale-up/scale-down simulation vs updating the expected state of the cluster) and not necessarily on the success / failure of the logic. Or, in other words, just reverting a forked snapshot and proceeding is generally not the error handling strategy for a failed simulation. By using `func() error` we have no convenient way of propagating the error outside of WithForkedSnapshot.

I'd much rather this took `func() (bool, error)` where the boolean return value determines if the forked snapshot should be commited or reverted. This would separate the decision from any other error handling and force the user to make explicit decision on whether the snapshot should be committed or reverted for each possible error condition.

Reply:
Agreed, thanks for pointing this out! I modified the function and changed the error returned from WithForkedSnapshot regarding cluster snapshot operation. I think we should return error that shows if expected operation was completed (Commit or Revert). 
However all of the implementations of Revert() return nil. And I suppose since Revert() is direct opposite to Fork(), so they both should return the same value. So I actually think that we can drop returned value for Revert() as we did it for Fork(). WDYT?

Reply:
I guess if the revert cannot fail in any implementation, I think we can drop the error from signature. I don't really expect implementations of this interface outside the ones we have, so changing interface doesn't seem like an issue.
The only condition I can think of when revert could fail is if the snapshot was not forked, but IIRC we just treat this as no-op and don't return an error?

Reply:
Yes, if we haven't fork before calling revert we just do nothing and return nil

Reply:
Dropped error for Revert()